builtin/credential/aws/path_role.go

```diff
@@ -320,6 +322,64 @@ func (b *backend) setRole(ctx context.Context, s logical.Storage, roleName strin
 	return nil
 }
 
+// updateUpgradableRoleEntries upgrades and persists all of the role entries
+// that are in need of being upgraded.
+func (b *backend) updateUpgradableRoleEntries(ctx context.Context, req *logical.InitializationRequest) error {
+
+	// Upgrade only if we are either: (1) a local mount, or (2) are _not_ a
+	// performance replicated standby cluster.
+	if !(b.System().LocalMount() || !b.System().ReplicationState().HasState(
+		consts.ReplicationPerformanceSecondary|consts.ReplicationPerformanceStandby)) {
+		return nil
+	}
+
+	// Read all the role names.
+	b.roleMutex.RLock()
```

Comment:
You don't need to grab the mutex here for listing. The reason is that if we're in the process of creating a new mount, it will already happen on the active node (since it's a write) and will properly use the latest role structure.

Reply:
I'd be happy to remove that lock, but I just want to double check -- `Initialize()` is being called ansynchronously from a `postUnseal` func in `Core.setupMounts()`, so I think maybe holding the read lock would be a good idea?  https://github.com/hashicorp/vault/blob/aws-upgrade-role-entries/vault/mount.go#L1119-L1126

Reply:
Those functions are actually called _before_ we start serving requests so there should be no other request running. (And even if there were, my point still stands :-) )

Reply:
ok :-), the `RLock()` is gone, and its a `Mutex` now